express-notes-api/index.js

```diff
@@ -0,0 +1,107 @@
+const express = require('express');
+const app = express();
+const data = require('./data.json')
+const notes = data.notes;
+const fs = require('fs');
+
+app.use(express.json());
+
+app.get('/api/notes', (req, res) => {
+  const arr = [];
+  for (const x in notes){
+    arr.push(notes[x]);
+  }
+  res.status(200).json(arr);
+})
+
+app.get('/api/notes/:id', (req, res) => {
+  const getId = req.params.id;
+  const id = Number(getId);
+  if(!Number.isInteger(id) || id <= 0){
+    const error400 = {error: "id must be a positive integer"}
+    res.status(400).json(error400);
+  }
+  else if(Number.isInteger(id) && id in notes) {
+    res.json(notes[id]);
+  }
+  else{
+    const error404 = {error: `cannot find note with id ${id}`};
+    res.status(404).json(error404);
+  }
+})
+
+const update = (index, newNote) => {
+  fs.readFile('./data.json', 'utf8', (err, data) => {
+    if (err) throw err;
+    notes[index] = newNote;
+    const newData = JSON.stringify(data, null, 2);
+    fs.writeFile('data.json', newData, 'utf8', (err) => {
+      if (err) throw err;
+    });
+  });
+}
+
+app.post('/api/notes', (req, res) => {
+  if(!req.body.content){
+    res.status(400).json({ "error": "content is a required field"})
+  }
+  else {
+    const newNote = {
+      id: data.nextId,
+      content: req.body.content
+    };
+    notes[data.nextId] = newNote;
+    update(data.nextId, notes[data.nextId]);
+    data.nextId++;
+    res.status(201).json(newNote)
+  }
```

Comment:
Because your update runs asynchronously, you are sending back a success response before the file system operation is complete. I would recommend moving the file operations to the individual endpoints, as the knowledge required to use a helper function like this to accomplish the goal are things we have not covered yet. But, a very clever idea for a solution!